artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java

```diff
@@ -289,9 +290,23 @@ public void applySetting(final AddressSettings addressSettings) {
          this.pageLimitBytes = null;
       }
 
+      this.pageLimitMessages = addressSettings.getPageLimitMessages();
+      if (firstTime || !Objects.equals(this.pageLimitMessages, originalLimitMessages)) {
+         if (this.cursorProvider != null) {
+            this.cursorProvider.checkClearPageLimit();
+         }
+      }
```

Comment:
The cursorProvider hasnt been created at the point this is called during the constructor, so its never going to be non-null if 'firstTime' is true and it cant be non-null in real usage if it is false. My suggestion on the prior round was actually that it should _not_ do this for the constructor, since this is handled via the start etc processes later after construction and so would basically be duplicate.

This is also going to check with the _previous_ estimated max pages rather than any new value, because that value hasnt been updated yet (its done later below). I dont think that makes sense, as makes it possible this may want to clear the limit, only for the very next bit to reinstate it because it should never have been cleared with the updated setting.

As I said, while applying the settings values, I would use a flag for each of pageLimitMessages and the estimated max pages, to remember whether they have changed....then at end, after application, check things when both are effective together.